File: internal/computeruse/action.go
// Package computeruse implements the device-scoped computer-use control
// surface and its Locked Use extension.
//
// Two boundaries matter here and are kept deliberately separate:
//
//   - Actions (this file) describe what the agent may do to the desktop. The
//     vocabulary is closed: an unknown action id is rejected rather than passed
//     through to a shell or a native helper.
//   - Locked Use (locked.go, grant.go) decides whether the desktop may be
//     unlocked at all while the screen is locked. It never handles the user's
//     password; it only mints short-lived signed grants that a separately
//     installed Apple Authorization Plug-in verifies during the unlock flow.
package computeruse

import (
	"errors"
	"fmt"
	"strings"
)

// ActionID is the closed set of desktop operations. Anything outside this list
// is refused at the API boundary, so a compromised or confused caller cannot
// reach an arbitrary command through the computer-use route.
type ActionID string

const (
	ActionScreenshot ActionID = "screen.capture"
	ActionMove       ActionID = "pointer.move"
	ActionClick      ActionID = "pointer.click"
	ActionScroll     ActionID = "pointer.scroll"
	ActionType       ActionID = "keyboard.type"
	ActionKey        ActionID = "keyboard.key"
)

// MouseButton is the closed set of pointer buttons.
type MouseButton string

const (
	ButtonLeft   MouseButton = "left"
	ButtonRight  MouseButton = "right"
	ButtonMiddle MouseButton = "middle"
)

const (
	// MaxTypeRunes bounds a single type action. Long text is the caller's job
	// to chunk; an unbounded synthetic keystroke stream is a denial-of-service
	// against the very desktop the agent is driving.
	MaxTypeRunes = 4096
	// MaxKeyChordKeys bounds a single key chord (e.g. cmd+shift+4).
	MaxKeyChordKeys = 5
	// MaxClickCount bounds a multi-click (single, double, triple).
	MaxClickCount = 3
	// MaxScrollMagnitude bounds one scroll action's wheel delta.
	MaxScrollMagnitude = 4096
)

// Action is one validated desktop operation. It is produced only by
// ParseAction, so every field a native helper consumes has already been
// range-checked.
type Action struct {
	ID     ActionID    `json:"id"`
	X      int         `json:"x,omitempty"`
	Y      int         `json:"y,omitempty"`
	Button MouseButton `json:"button,omitempty"`
	Count  int         `json:"count,omitempty"`
	Text   string      `json:"text,omitempty"`
	Keys   []string    `json:"keys,omitempty"`
	DeltaX int         `json:"delta_x,omitempty"`
	DeltaY int         `json:"delta_y,omitempty"`
}

// ActionRequest is the raw wire shape accepted by the API. It is intentionally
// distinct from Action: the API never hands an unvalidated struct to the
// system layer.
type ActionRequest struct {
	Action string   `json:"action"`
	X      *int     `json:"x"`
	Y      *int     `json:"y"`
	Button string   `json:"button"`
	Count  int      `json:"count"`
	Text   string   `json:"text"`
	Keys   []string `json:"keys"`
	DeltaX int      `json:"delta_x"`
	DeltaY int      `json:"delta_y"`
}

// keyNames is the closed set of non-character key names a chord may use.
// Character keys (a-z, 0-9, punctuation) are handled as single-rune names.
var keyNames = map[string]bool{
	"cmd": true, "command": true, "ctrl": true, "control": true,
	"alt": true, "option": true, "shift": true, "fn": true,
	"return": true, "enter": true, "tab": true, "space": true,
	"delete": true, "backspace": true, "escape": true, "esc": true,
	"up": true, "down": true, "left": true, "right": true,
	"home": true, "end": true, "pageup": true, "pagedown": true,
	"f1": true, "f2": true, "f3": true, "f4": true, "f5": true, "f6": true,
	"f7": true, "f8": true, "f9": true, "f10": true, "f11": true, "f12": true,
}

var errUnknownAction = errors.New("unknown computer-use action")

// ParseAction validates a wire request into an Action. It is the only way to
// construct an Action that the system layer will execute.
func ParseAction(in ActionRequest) (Action, error) {
	id := ActionID(strings.TrimSpace(in.Action))
	switch id {
	case ActionScreenshot:
		return Action{ID: id}, nil

	case ActionMove:
		x, y, err := requirePoint(in)
		if err != nil {
			return Action{}, err
		}
		return Action{ID: id, X: x, Y: y}, nil

	case ActionClick:
		x, y, err := requirePoint(in)
		if err != nil {
			return Action{}, err
		}
		button, err := parseButton(in.Button)
		if err != nil {
			return Action{}, err
		}
		count := in.Count
		if count == 0 {
			count = 1
		}
		if count < 1 || count > MaxClickCount {
			return Action{}, fmt.Errorf("click count must be 1..%d", MaxClickCount)
		}
		return Action{ID: id, X: x, Y: y, Button: button, Count: count}, nil

	case ActionScroll:
		x, y, err := requirePoint(in)
		if err != nil {
			return Action{}, err
		}
		if in.DeltaX == 0 && in.DeltaY == 0 {
			return Action{}, errors.New("scroll requires a non-zero delta_x or delta_y")
		}
		if abs(in.DeltaX) > MaxScrollMagnitude || abs(in.DeltaY) > MaxScrollMagnitude {
			return Action{}, fmt.Errorf("scroll delta must be within +/-%d", MaxScrollMagnitude)
		}
		return Action{ID: id, X: x, Y: y, DeltaX: in.DeltaX, DeltaY: in.DeltaY}, nil

	case ActionType:
		if in.Text == "" {
			return Action{}, errors.New("type requires text")
		}
		if n := len([]rune(in.Text)); n > MaxTypeRunes {
			return Action{}, fmt.Errorf("type text exceeds %d characters", MaxTypeRunes)
		}
		if strings.ContainsRune(in.Text, 0) {
			return Action{}, errors.New("type text contains a NUL byte")
		}
		return Action{ID: id, Text: in.Text}, nil

	case ActionKey:
		keys, err := parseKeys(in.Keys)
		if err != nil {
			return Action{}, err
		}
		return Action{ID: id, Keys: keys}, nil

	default:
		return Action{}, errUnknownAction
	}
}

func requirePoint(in ActionRequest) (int, int, error) {
	if in.X == nil || in.Y == nil {
		return 0, 0, errors.New("action requires x and y")
	}
	// Screen bounds are enforced by the native helper against the real display
	// geometry. Reject only absurd values here so a typo cannot become a
	// pathological synthetic event.
	if *in.X < -32768 || *in.X > 32767 || *in.Y < -32768 || *in.Y > 32767 {
		return 0, 0, errors.New("coordinates out of range")
	}
	return *in.X, *in.Y, nil
}

func parseButton(name string) (MouseButton, error) {
	switch strings.ToLower(strings.TrimSpace(name)) {
	case "", "left":
		return ButtonLeft, nil
	case "right":
		return ButtonRight, nil
	case "middle":
		return ButtonMiddle, nil
	default:
		return "", errors.New("unknown mouse button: " + name)
	}
}

func parseKeys(in []string) ([]string, error) {
	if len(in) == 0 {
		return nil, errors.New("key action requires keys")
	}
	if len(in) > MaxKeyChordKeys {
		return nil, fmt.Errorf("key chord exceeds %d keys", MaxKeyChordKeys)
	}
	out := make([]string, 0, len(in))
	for _, raw := range in {
		key := strings.ToLower(strings.TrimSpace(raw))
		if key == "" {
			return nil, errors.New("key chord contains an empty key")
		}
		if !keyNames[key] && len([]rune(key)) != 1 {
			return nil, errors.New("unknown key: " + raw)
		}
		out = append(out, key)
	}
	return out, nil
}

func abs(v int) int {
	if v < 0 {
		return -v
	}
	return v
}

File: mac/RemoteAgentDesktop/Sources/AgentHaloDesktopCore/Action.swift
import Foundation

/// The closed set of desktop operations. Anything outside this list is refused
/// before it reaches CoreGraphics, so a confused or compromised caller cannot
/// reach an arbitrary command through this surface.
///
/// This is a port of internal/computeruse/action.go and must stay behaviourally
/// identical to it while the Go side still validates: two validators that
/// disagree would mean the reachable vocabulary is the union of both, which is
/// larger than either was reviewed to be.
public enum ActionID: String, Codable, CaseIterable, Sendable {
    case screenCapture = "screen.capture"
    case pointerMove = "pointer.move"
    case pointerClick = "pointer.click"
    case pointerScroll = "pointer.scroll"
    case keyboardType = "keyboard.type"
    case keyboardKey = "keyboard.key"
}

public enum MouseButton: String, Codable, Sendable {
    case left, right, middle
}

/// Pointer coordinates coming from public/debug APIs retain the historical
/// Core Graphics global-coordinate contract. Model tools instead point into
/// the top-left-origin composite PNG returned by `get_app_state`; the desktop
/// layer translates those coordinates across negative-origin displays.
public enum ActionCoordinateSpace: String, Codable, Sendable {
    case global
    case screenshot
}

public enum ActionLimits {
    /// Bounds a single type action. Long text is the caller's job to chunk; an
    /// unbounded synthetic keystroke stream is a denial-of-service against the
    /// very desktop the agent is driving.
    public static let maxTypeCharacters = 4096
    /// Bounds a single chord (e.g. cmd+shift+4).
    public static let maxChordKeys = 5
    /// Bounds a multi-click (single, double, triple).
    public static let maxClickCount = 3
    /// Bounds one scroll action's wheel delta.
    public static let maxScrollMagnitude = 4096
    /// Screen bounds are enforced against real display geometry when the event
    /// is posted. This only rejects absurd values, so a typo cannot become a
    /// pathological synthetic event.
    public static let coordinateRange = (-32768)...32767
}

/// One validated desktop operation. It is produced only by `Action.parse`, so
/// every field the CoreGraphics layer consumes has already been range-checked.
public struct Action: Sendable, Equatable {
    public let id: ActionID
    public let coordinateSpace: ActionCoordinateSpace
    public let x: Int
    public let y: Int
    public let button: MouseButton
    public let count: Int
    public let text: String
    public let keys: [String]
    public let deltaX: Int
    public let deltaY: Int

    init(
        id: ActionID, coordinateSpace: ActionCoordinateSpace = .global,
        x: Int = 0, y: Int = 0, button: MouseButton = .left,
        count: Int = 1, text: String = "", keys: [String] = [],
        deltaX: Int = 0, deltaY: Int = 0
    ) {
        self.id = id
        self.coordinateSpace = coordinateSpace
        self.x = x
        self.y = y
        self.button = button
        self.count = count
        self.text = text
        self.keys = keys
        self.deltaX = deltaX
        self.deltaY = deltaY
    }
}

/// The raw wire shape. Deliberately distinct from `Action`: nothing hands an
/// unvalidated request to the system layer.
public struct ActionRequest: Codable, Sendable {
    public var action: String
    public var coordinateSpace: String?
    public var x: Int?
    public var y: Int?
    public var button: String?
    public var count: Int?
    public var text: String?
    public var keys: [String]?
    public var deltaX: Int?
    public var deltaY: Int?

    enum CodingKeys: String, CodingKey {
        case action, x, y, button, count, text, keys
        case coordinateSpace = "coordinate_space"
        case deltaX = "delta_x"
        case deltaY = "delta_y"
    }

    public init(
        action: String, coordinateSpace: String? = nil,
        x: Int? = nil, y: Int? = nil, button: String? = nil,
        count: Int? = nil, text: String? = nil, keys: [String]? = nil,
        deltaX: Int? = nil, deltaY: Int? = nil
    ) {
        self.action = action
        self.coordinateSpace = coordinateSpace
        self.x = x
        self.y = y
        self.button = button
        self.count = count
        self.text = text
        self.keys = keys
        self.deltaX = deltaX
        self.deltaY = deltaY
    }
}

public struct ActionError: Error, Equatable, CustomStringConvertible {
    public let message: String
    public var description: String { message }
    init(_ message: String) { self.message = message }
}

/// The closed set of non-character key names a chord may use. Character keys
/// (a-z, 0-9, punctuation) are accepted as single-character names.
private let namedKeys: Set<String> = [
    "cmd", "command", "ctrl", "control", "alt", "option", "shift", "fn",
    "return", "enter", "tab", "space", "delete", "backspace", "escape", "esc",
    "up", "down", "left", "right", "home", "end", "pageup", "pagedown",
    "f1", "f2", "f3", "f4", "f5", "f6", "f7", "f8", "f9", "f10", "f11", "f12",
]

extension Action {
    /// Validates a wire request into an Action. This is the only constructor
    /// the system layer will execute.
    public static func parse(_ request: ActionRequest) throws -> Action {
        guard let id = ActionID(rawValue: request.action.trimmingCharacters(in: .whitespaces)) else {
            throw ActionError("unknown computer-use action")
        }
        switch id {
        case .screenCapture:
            return Action(id: id)

        case .pointerMove:
            let (space, x, y) = try requirePoint(request)
            return Action(id: id, coordinateSpace: space, x: x, y: y)

        case .pointerClick:
            let (space, x, y) = try requirePoint(request)
            let button = try parseButton(request.button)
            let count = request.count ?? 1
            guard count >= 1, count <= ActionLimits.maxClickCount else {
                throw ActionError("click count must be 1..\(ActionLimits.maxClickCount)")
            }
            return Action(
                id: id, coordinateSpace: space, x: x, y: y,
                button: button, count: count)

        case .pointerScroll:
            let (space, x, y) = try requirePoint(request)
            let dx = request.deltaX ?? 0
            let dy = request.deltaY ?? 0
            guard dx != 0 || dy != 0 else {
                throw ActionError("scroll requires a non-zero delta_x or delta_y")
            }
            guard abs(dx) <= ActionLimits.maxScrollMagnitude,
                  abs(dy) <= ActionLimits.maxScrollMagnitude else {
                throw ActionError("scroll delta must be within +/-\(ActionLimits.maxScrollMagnitude)")
            }
            return Action(
                id: id, coordinateSpace: space, x: x, y: y,
                deltaX: dx, deltaY: dy)

        case .keyboardType:
            let text = request.text ?? ""
            guard !text.isEmpty else { throw ActionError("type requires text") }
            guard text.count <= ActionLimits.maxTypeCharacters else {
                throw ActionError("type text exceeds \(ActionLimits.maxTypeCharacters) characters")
            }
            guard !text.unicodeScalars.contains(where: { $0.value == 0 }) else {
                throw ActionError("type text contains a NUL byte")
            }
            return Action(id: id, text: text)

        case .keyboardKey:
            return Action(id: id, keys: try parseKeys(request.keys ?? []))
        }
    }

    private static func requirePoint(
        _ request: ActionRequest
    ) throws -> (ActionCoordinateSpace, Int, Int) {
        guard let x = request.x, let y = request.y else {
            throw ActionError("action requires x and y")
        }
        guard ActionLimits.coordinateRange.contains(x),
              ActionLimits.coordinateRange.contains(y) else {
            throw ActionError("coordinates out of range")
        }
        let rawSpace = (request.coordinateSpace ?? "global")
            .trimmingCharacters(in: .whitespaces).lowercased()
        guard let space = ActionCoordinateSpace(rawValue: rawSpace) else {
            throw ActionError("unknown coordinate space")
        }
        if space == .screenshot, (x < 0 || y < 0) {
            throw ActionError("screenshot coordinates cannot be negative")
        }
        return (space, x, y)
    }

    private static func parseButton(_ name: String?) throws -> MouseButton {
        let cleaned = (name ?? "").trimmingCharacters(in: .whitespaces).lowercased()
        if cleaned.isEmpty { return .left }
        guard let button = MouseButton(rawValue: cleaned) else {
            throw ActionError("unknown mouse button: \(name ?? "")")
        }
        return button
    }

    private static func parseKeys(_ keys: [String]) throws -> [String] {
        guard !keys.isEmpty else { throw ActionError("key action requires keys") }
        guard keys.count <= ActionLimits.maxChordKeys else {
            throw ActionError("key chord exceeds \(ActionLimits.maxChordKeys) keys")
        }
        return try keys.map { raw in
            let key = raw.trimmingCharacters(in: .whitespaces).lowercased()
            guard !key.isEmpty else { throw ActionError("key chord contains an empty key") }
            guard namedKeys.contains(key) || key.count == 1 else {
                throw ActionError("unknown key: \(raw)")
            }
            return key
        }
    }
}
